File: src/com/lushprojects/circuitjs1/client/TriStateElm.java
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/



package com.lushprojects.circuitjs1.client;

import com.google.gwt.xml.client.Element;
import com.google.gwt.xml.client.Document;

// contributed by Edward Calver

class TriStateElm extends CircuitElm {
    double resistance, r_on, r_off, r_off_ground, highVoltage;
    int busWidth = 1;
    VoltageSource voltageSources[];

    // Unfortunately we need all three flags to keep track of flipping.
    // FLAG_FLIP_X/Y affect the rounding direction if the elm is an odd grid length.
    // FLAG_FLIP does not.
    final int FLAG_FLIP = 1;
    final int FLAG_FLIP_X = 2;
    final int FLAG_FLIP_Y = 4;

    public TriStateElm(int xx, int yy) {
	super(xx, yy);
	r_on = 0.1;
	r_off = 1e10;
	r_off_ground = 1e8;
	noDiagonal = true;

        // copy defaults from last gate edited
        highVoltage = GateElm.lastHighVoltage;
    }

    public TriStateElm(int xa, int ya, int xb, int yb, int f, StringTokenizer st) {
	super(xa, ya, xb, yb, f);
	r_on = 0.1;
	r_off = 1e10;
	r_off_ground = 0;
	noDiagonal = true;
	highVoltage = 5;
	try {
	    r_on = new Double(st.nextToken()).doubleValue();
	    r_off = new Double(st.nextToken()).doubleValue();
	    r_off_ground = new Double(st.nextToken()).doubleValue();
            highVoltage = new Double (st.nextToken()).doubleValue();
	} catch (Exception e) {
	}

    }

    String dump() {
	return super.dump() + " " + r_on + " " + r_off + " " + r_off_ground + " " + highVoltage;
    }

    void dumpXml(Document doc, Element elem) {
        super.dumpXml(doc, elem);
        XMLSerializer.dumpAttr(elem, "ron", r_on);
        XMLSerializer.dumpAttr(elem, "roff", r_off);
        XMLSerializer.dumpAttr(elem, "rog", r_off_ground);
        XMLSerializer.dumpAttr(elem, "hi", highVoltage);
	if (busWidth != 1)
	    XMLSerializer.dumpAttr(elem, "bw", busWidth);
    }

    void undumpXml(XMLDeserializer xml) {
        super.undumpXml(xml);
        r_on = xml.parseDoubleAttr("ron", r_on);
        r_off = xml.parseDoubleAttr("roff", r_off);
        r_off_ground = xml.parseDoubleAttr("rog", r_off_ground);
        highVoltage = xml.parseDoubleAttr("hi", highVoltage);
	busWidth = xml.parseIntAttr("bw", 1);
    }

    int getDumpType() { return 180; }
    String getXmlDumpType() { return "ts"; }
    void setHighVoltage(double hv) { highVoltage = hv; }

    boolean open;

    Point ps, point3, lead3, busLead1;

    Polygon gatePoly;

    void setPoints() {
	super.setPoints();
	int len = 32;
	calcLeads(len);
	adjustLeadsToGrid((flags & FLAG_FLIP_X) != 0, (flags & FLAG_FLIP_Y) != 0);

	ps = new Point();
	int hs = 16;

	int ww = 16;
	if (ww > dn / 2)
	    ww = (int) (dn / 2);
	Point triPoints[] = newPointArray(3);
	interpPoint2(lead1, lead2, triPoints[0], triPoints[1], 0, hs + 2);
	triPoints[2] = interpPoint(lead1, lead2, .5 + (ww - 2) / (double)len);
	gatePoly = createPolygon(triPoints);

	// busLead1 is lead1 pulled back slightly so thick bus line doesn't bleed into triangle
	busLead1 = interpPoint(point1, lead1, 1 - 2 / dn);

	int sign = ((flags & FLAG_FLIP) == 0) ? -1 : 1;
	point3 = interpPoint(lead1, lead2, .5, sign*hs);
	lead3 = interpPoint(lead1, lead2, .5, sign*(hs/2 + 2));
    }

    void draw(Graphics g) {
	int hs = 16;
	setBbox(point1, point2, hs);

	// draw control lead underneath the triangle
	setVoltageColor(g, volts[2 * busWidth]);
	drawThickLine(g, point3, lead3);

	// draw leads with bus thickness if bus mode
	if (busWidth > 1) {
	    setVoltageColor(g, volts[0]);
	    drawThickLine(g, point1, busLead1, 5);
	    setVoltageColor(g, volts[busWidth]);
	    drawThickLine(g, lead2, point2, 5);
	} else {
	    draw2Leads(g);
	}

	g.setColor(lightGrayColor);
	drawThickPolygon(g, gatePoly);
	curcount = updateDotCount(current, curcount);
	drawDots(g, lead2, point2, curcount);
	drawPosts(g);
    }

    // node layout:
    // nodes[0..busWidth-1]: input bus
    // nodes[busWidth..2*busWidth-1]: output bus
    // nodes[2*busWidth]: control
    // nodes[2*busWidth+1..3*busWidth]: internal nodes (one per bit)

    int controlNode() { return 2 * busWidth; }
    int internalNode(int bit) { return 2 * busWidth + 1 + bit; }

    void calculateCurrent() {
	current = 0;
	for (int i = 0; i < busWidth; i++) {
	    int intNode = internalNode(i);
	    int outNode = busWidth + i;
	    double current31 = (volts[intNode] - volts[outNode]) / resistance;
	    double current10 = (r_off_ground == 0) ? 0 : volts[outNode] / r_off_ground;
	    current += current31 - current10;
	}
    }

    double getCurrentIntoNode(int n) {
	if (n >= busWidth && n < 2 * busWidth)
	    return current / busWidth;
	return 0;
    }

    // we need this to be able to change the matrix for each step
    boolean nonLinear() {
	return true;
    }

    void stamp() {
	for (int i = 0; i < busWidth; i++) {
	    sim.stampVoltageSource(CircuitNode.ground, nodes[internalNode(i)], voltageSources[i]);
	    sim.stampNonLinear(nodes[internalNode(i)]);
	    sim.stampNonLinear(nodes[busWidth + i]);
	}
    }

    void doStep() {
	open = (volts[controlNode()] < highVoltage * .5);
	resistance = (open) ? r_off : r_on;
	for (int i = 0; i < busWidth; i++) {
	    int intNode = internalNode(i);
	    int outNode = busWidth + i;
	    int inNode = i;
	    sim.stampResistor(nodes[intNode], nodes[outNode], resistance);

	    if (r_off_ground > 0)
		sim.stampResistor(nodes[outNode], CircuitNode.ground, r_off_ground);

	    sim.updateVoltageSource(CircuitNode.ground, nodes[intNode], voltageSources[i],
		volts[inNode] > highVoltage * .5 ? highVoltage : 0);
	}
    }

    void drag(int xx, int yy) {
	// use mouse to select which side the buffer enable should be on
	boolean flip = (xx < x) == (yy < y);

	xx = snapGrid(xx);
	yy = snapGrid(yy);
	if (abs(x - xx) < abs(y - yy))
	    xx = x;
	else {
	    flip = !flip;
	    yy = y;
	}
	flags = flip ? (flags | FLAG_FLIP) : (flags & ~FLAG_FLIP);
	super.drag(xx, yy);
    }

    // posts: busWidth inputs + busWidth outputs + 1 control
    int getPostCount() {
	return 2 * busWidth + 1;
    }

    int getInternalNodeCount() {
	return busWidth;
    }

    int getVoltageSourceCount() {
	return busWidth;
    }

    void setVoltageSource(int n, VoltageSource v) {
	if (voltageSources == null || voltageSources.length != busWidth)
	    voltageSources = new VoltageSource[busWidth];
	voltageSources[n] = v;
	v.setNodes(CircuitNode.ground, nodes[internalNode(n)]);
    }

    boolean getMatrixConnection(int n1, int n2) {
	// each internal node connects to its corresponding output node
	for (int i = 0; i < busWidth; i++)
	    if (comparePair(n1, n2, busWidth + i, internalNode(i)))
		return true;
	return false;
    }

    Point getPost(int n) {
	if (n < busWidth)
	    return (busWidth > 1) ? new Point(point1.x, point1.y, n) : point1;
	if (n < 2 * busWidth)
	    return (busWidth > 1) ? new Point(point2.x, point2.y, n - busWidth) : point2;
	return point3;
    }

    int getPostWidth(int n) {
	if (n < 2 * busWidth)
	    return busWidth;
	return 1;
    }

    void getInfo(String arr[]) {
	arr[0] = "tri-state buffer";
	if (busWidth > 1)
	    arr[0] += " (" + busWidth + ")";
	arr[1] = open ? "open" : "closed";
	arr[2] = "Vd = " + getVoltageDText(getVoltageDiff());
	arr[3] = "I = " + getCurrentDText(getCurrent());
	arr[4] = "Vc = " + getVoltageText(volts[controlNode()]);
    }

    // there is no current path through the input, but there
    // is an indirect path through the output to ground.
    boolean getConnection(int n1, int n2) {
	return false;
    }

    boolean hasGroundConnection(int n1) {
	return (n1 >= busWidth && n1 < 2 * busWidth);
    }

    public EditInfo getEditInfo(int n) {
	if (n == 0)
	    return new EditInfo("On Resistance (ohms)", r_on, 0, 0).setPositive();
	if (n == 1)
	    return new EditInfo("Off Resistance (ohms)", r_off, 0, 0).setPositive();
	if (n == 2)
	    return new EditInfo("Output Pulldown Resistance (ohms)", r_off_ground, 0, 0).setPositive();
        if (n == 3)
            return new EditInfo("High Logic Voltage", highVoltage, 1, 10);
	if (n == 4)
	    return new EditInfo("Bus Width", busWidth, 1, 32).setDimensionless();
	return null;
    }

    public void setEditValue(int n, EditInfo ei) {

	if (n == 0 && ei.value > 0)
	    r_on = ei.value;
	if (n == 1 && ei.value > 0)
	    r_off = ei.value;
	if (n == 2 && ei.value > 0)
	    r_off_ground = ei.value;
	if (n == 3)
            highVoltage = GateElm.lastHighVoltage = ei.value;
	if (n == 4) {
	    if (ei.value >= 1) {
		busWidth = (int) ei.value;
		allocNodes();
	    } else
		ei.setError("must be >= 1");
	}
    }

    void flipX(int c2, int count) {
	flags ^= FLAG_FLIP|FLAG_FLIP_X;
	super.flipX(c2, count);
    }

    void flipY(int c2, int count) {
	flags ^= FLAG_FLIP|FLAG_FLIP_Y;
	super.flipY(c2, count);
    }

    void flipXY(int c2, int count) {
	flags ^= FLAG_FLIP;
	super.flipXY(c2, count);
    }
}

File: TriStateElm.ts
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

// contributed by Edward Calver

import { CircuitElm } from "./CircuitElm";
import { CircuitNode } from "./CircuitNode";
import { CircuitXMLSerializer } from "./CircuitXMLSerializer";
import { CircuitXMLDeserializer } from "./CircuitXMLDeserializer";
import { EditInfo } from "./EditInfo";
import { GateElm } from "./GateElm";
import { Graphics } from "./Graphics";
import { Point } from "./Point";
import { Polygon } from "./Polygon";
import { StringTokenizer } from "./StringTokenizer";
import { VoltageSource } from "./VoltageSource";
import { parseFloatStrict } from "./NumberParse";

export class TriStateElm extends CircuitElm {
    resistance: number = 0;
    r_on: number;
    r_off: number;
    r_off_ground: number;
    highVoltage: number;
    busWidth: number = 1;
    voltageSources: VoltageSource[];

    readonly FLAG_FLIP   = 1;
    readonly FLAG_FLIP_X = 2;
    readonly FLAG_FLIP_Y = 4;

    open: boolean = false;
    ps: Point;
    point3: Point;
    lead3: Point;
    busLead1: Point;
    gatePoly: Polygon;

    constructor(xx: number, yy: number);
    constructor(xa: number, ya: number, xb: number, yb: number, f: number, st: StringTokenizer);
    constructor(xa: number, ya: number, xb?: number, yb?: number, f?: number, st?: StringTokenizer) {
        super(xa, ya, xb, yb, f);
        if (st !== undefined) {
            this.r_on = 0.1; this.r_off = 1e10; this.r_off_ground = 0; this.highVoltage = 5;
            try {
                this.r_on = parseFloatStrict(st.nextToken());
                this.r_off = parseFloatStrict(st.nextToken());
                this.r_off_ground = parseFloatStrict(st.nextToken());
                this.highVoltage = parseFloatStrict(st.nextToken());
            } catch (e) {}
        } else {
            this.r_on = 0.1; this.r_off = 1e10; this.r_off_ground = 1e8;
            this.highVoltage = GateElm.lastHighVoltage;
        }
        this.noDiagonal = true;
        this.allocNodes();
    }

    dump(): string {
        return super.dump() + " " + this.r_on + " " + this.r_off + " " + this.r_off_ground + " " + this.highVoltage;
    }

    dumpXml(doc: Document, elem: Element): void {
        super.dumpXml(doc, elem);
        CircuitXMLSerializer.dumpAttr(elem, "ron", this.r_on);
        CircuitXMLSerializer.dumpAttr(elem, "roff", this.r_off);
        CircuitXMLSerializer.dumpAttr(elem, "rog", this.r_off_ground);
        CircuitXMLSerializer.dumpAttr(elem, "hi", this.highVoltage);
        if (this.busWidth !== 1)
            CircuitXMLSerializer.dumpAttr(elem, "bw", this.busWidth);
    }

    undumpXml(xml: CircuitXMLDeserializer): void {
        super.undumpXml(xml);
        this.r_on = xml.parseDoubleAttr("ron", this.r_on);
        this.r_off = xml.parseDoubleAttr("roff", this.r_off);
        this.r_off_ground = xml.parseDoubleAttr("rog", this.r_off_ground);
        this.highVoltage = xml.parseDoubleAttr("hi", this.highVoltage);
        this.busWidth = xml.parseIntAttr("bw", 1);
        this.allocNodes();
    }

    getDumpType(): number { return 180; }
    getXmlDumpType(): string { return "ts"; }
    setHighVoltage(hv: number): void { this.highVoltage = hv; }

    setPoints(): void {
        super.setPoints();
        const len = 32;
        this.calcLeads(len);
        this.adjustLeadsToGrid((this.flags & this.FLAG_FLIP_X) !== 0, (this.flags & this.FLAG_FLIP_Y) !== 0);
        this.ps = new Point();
        const hs = 16;
        let ww = 16;
        if (ww > this.dn / 2) ww = Math.trunc(this.dn / 2);
        const triPoints = this.newPointArray(3);
        this.interpPoint2(this.lead1!, this.lead2!, triPoints[0], triPoints[1], 0, hs + 2);
        triPoints[2] = this.interpPoint(this.lead1!, this.lead2!, 0.5 + (ww - 2) / len);
        this.gatePoly = this.createPolygon(triPoints);
        this.busLead1 = this.interpPoint(this.point1, this.lead1!, 1 - 2 / this.dn);
        const sign = (this.flags & this.FLAG_FLIP) === 0 ? -1 : 1;
        this.point3 = this.interpPoint(this.lead1!, this.lead2!, 0.5, sign * hs);
        this.lead3  = this.interpPoint(this.lead1!, this.lead2!, 0.5, sign * (hs/2 + 2));
    }

    draw(g: Graphics): void {
        const hs = 16;
        this.setBbox(this.point1, this.point2, hs);
        this.setVoltageColor(g, this.nodes[2 * this.busWidth].v);
        CircuitElm.drawThickLine(g, this.point3, this.lead3);
        if (this.busWidth > 1) {
            this.setVoltageColor(g, this.nodes[0].v);
            CircuitElm.drawThickLine(g, this.point1, this.busLead1, 5);
            this.setVoltageColor(g, this.nodes[this.busWidth].v);
            CircuitElm.drawThickLine(g, this.lead2!, this.point2, 5);
        } else {
            this.draw2Leads(g);
        }
        g.setColor(CircuitElm.lightGrayColor);
        CircuitElm.drawThickPolygon(g, this.gatePoly);
        this.curcount = this.updateDotCountImpl(this.current, this.curcount);
        this.drawDots(g, this.lead2!, this.point2, this.curcount);
        this.drawPosts(g);
    }

    controlNode(): number { return 2 * this.busWidth; }
    internalNode(bit: number): number { return 2 * this.busWidth + 1 + bit; }

    calculateCurrent(): void {
        this.current = 0;
        for (let i = 0; i < this.busWidth; i++) {
            const intNode = this.internalNode(i);
            const outNode = this.busWidth + i;
            const current31 = (this.nodes[intNode].v - this.nodes[outNode].v) / this.resistance;
            const current10 = this.r_off_ground === 0 ? 0 : this.nodes[outNode].v / this.r_off_ground;
            this.current += current31 - current10;
        }
    }

    getCurrentIntoNode(n: number): number {
        if (n >= this.busWidth && n < 2 * this.busWidth) return this.current / this.busWidth;
        return 0;
    }

    nonLinear(): boolean { return true; }

    stamp(): void {
        for (let i = 0; i < this.busWidth; i++) {
            CircuitElm.sim.stampVoltageSource(CircuitNode.ground, this.nodes[this.internalNode(i)], this.voltageSources[i]);
            CircuitElm.sim.stampNonLinear(this.nodes[this.internalNode(i)]);
            CircuitElm.sim.stampNonLinear(this.nodes[this.busWidth + i]);
        }
    }

    doStep(): void {
        this.open = this.nodes[this.controlNode()].v < this.highVoltage * 0.5;
        this.resistance = this.open ? this.r_off : this.r_on;
        for (let i = 0; i < this.busWidth; i++) {
            const intNode = this.internalNode(i);
            const outNode = this.busWidth + i;
            CircuitElm.sim.stampResistor(this.nodes[intNode], this.nodes[outNode], this.resistance);
            if (this.r_off_ground > 0)
                CircuitElm.sim.stampResistor(this.nodes[outNode], CircuitNode.ground, this.r_off_ground);
            CircuitElm.sim.updateVoltageSource(CircuitNode.ground, this.nodes[intNode], this.voltageSources[i],
                this.nodes[i].v > this.highVoltage * 0.5 ? this.highVoltage : 0);
        }
    }

    drag(xx: number, yy: number): void {
        let flip = (xx < this.x) === (yy < this.y);
        xx = this.snapGrid(xx); yy = this.snapGrid(yy);
        if (Math.abs(this.x - xx) < Math.abs(this.y - yy)) xx = this.x;
        else { flip = !flip; yy = this.y; }
        this.flags = flip ? (this.flags | this.FLAG_FLIP) : (this.flags & ~this.FLAG_FLIP);
        super.drag(xx, yy);
    }

    getPostCount(): number { return 2 * this.busWidth + 1; }
    getInternalNodeCount(): number { return this.busWidth; }
    getVoltageSourceCount(): number { return this.busWidth; }

    setVoltageSource(n: number, v: VoltageSource): void {
        if (!this.voltageSources || this.voltageSources.length !== this.busWidth)
            this.voltageSources = new Array(this.busWidth);
        this.voltageSources[n] = v;
        v.setNodes(CircuitNode.ground, this.nodes[this.internalNode(n)]);
    }

    getMatrixConnection(n1: number, n2: number): boolean {
        for (let i = 0; i < this.busWidth; i++)
            if (this.comparePair(n1, n2, this.busWidth + i, this.internalNode(i))) return true;
        return false;
    }

    getPost(n: number): Point {
        if (n < this.busWidth)
            return this.busWidth > 1 ? new Point(this.point1.x, this.point1.y, n) : this.point1;
        if (n < 2 * this.busWidth)
            return this.busWidth > 1 ? new Point(this.point2.x, this.point2.y, n - this.busWidth) : this.point2;
        return this.point3;
    }

    getPostWidth(n: number): number { return n < 2 * this.busWidth ? this.busWidth : 1; }

    getInfo(arr: string[]): void {
        arr[0] = "tri-state buffer";
        if (this.busWidth > 1) arr[0] += " (" + this.busWidth + ")";
        arr[1] = this.open ? "open" : "closed";
        arr[2] = "Vd = " + CircuitElm.getVoltageDText(this.getVoltageDiff());
        arr[3] = "I = " + CircuitElm.getCurrentDText(this.getCurrent());
        arr[4] = "Vc = " + CircuitElm.getVoltageText(this.nodes[this.controlNode()].v);
    }

    getConnection(n1: number, n2: number): boolean { return false; }
    hasGroundConnection(n1: number): boolean { return n1 >= this.busWidth && n1 < 2 * this.busWidth; }

    getEditInfo(n: number): EditInfo | null {
        if (n === 0) return new EditInfo("On Resistance (ohms)", this.r_on, 0, 0).setPositive();
        if (n === 1) return new EditInfo("Off Resistance (ohms)", this.r_off, 0, 0).setPositive();
        if (n === 2) return new EditInfo("Output Pulldown Resistance (ohms)", this.r_off_ground, 0, 0).setPositive();
        if (n === 3) return new EditInfo("High Logic Voltage", this.highVoltage, 1, 10).setUnitStep();
        if (n === 4) return new EditInfo("Bus Width", this.busWidth, 1, 32).setDimensionless();
        return null;
    }

    setEditValue(n: number, ei: EditInfo): void {
        if (n === 0 && ei.value > 0) this.r_on = ei.value;
        if (n === 1 && ei.value > 0) this.r_off = ei.value;
        if (n === 2 && ei.value > 0) this.r_off_ground = ei.value;
        if (n === 3) this.highVoltage = GateElm.lastHighVoltage = ei.value;
        if (n === 4) {
            if (ei.value >= 1) { this.busWidth = Math.trunc(ei.value); this.allocNodes(); }
            else ei.setError("must be >= 1");
        }
    }

    flipX(c2: number, count: number): void { this.flags ^= this.FLAG_FLIP | this.FLAG_FLIP_X; super.flipX(c2, count); }
    flipY(c2: number, count: number): void { this.flags ^= this.FLAG_FLIP | this.FLAG_FLIP_Y; super.flipY(c2, count); }
    flipXY(c2: number, count: number): void { this.flags ^= this.FLAG_FLIP; super.flipXY(c2, count); }
}
